src/Disciplina/Educator/SizedMerkleTree.hs

```diff
@@ -0,0 +1,237 @@
+    go len xs = mkBranch (go i l) (go (len - i) r)
+      where
+        i = powerOfTwo len
+        (l, r) = splitAt i xs
+
+-- | Returns root of merkle tree.
+getSizedMerkleRoot :: SizedMerkleTree a -> SizedMerkleRoot a
+getSizedMerkleRoot SizedMerkleEmpty    = emptyHash
+getSizedMerkleRoot (SizedMerkleTree x) = mRoot x
+
+emptyHash :: SizedMerkleRoot a
+emptyHash = SizedMerkleRoot (hash mempty) 0
+
+-- | Return the largest power of two such that it's smaller than X.
+--
+-- >>> powerOfTwo 64
+-- 32
+-- >>> powerOfTwo 65
+-- 64
+powerOfTwo :: (Bits a, Num a) => a -> a
+powerOfTwo n
+    | n .&. (n - 1) == 0 = n `shiftR` 1
+    | otherwise = go n
+ where
+    {- “x .&. (x - 1)” clears the least significant bit:
+           ↓
+       01101000     x
+       01100111     x - 1
+       --------
+       01100000     x .&. (x - 1)
+     -}
+    go w = if w .&. (w - 1) == 0 then w else go (w .&. (w - 1))
+
+data MerkleProof a =
+  MerkleProof { getMerkleProof :: [ProofElem a]  -- ^ list of proof elements
+              , getMerkleProofRoot :: Maybe (SizedMerkleRoot a) -- ^ leaf root proof is
+                                                                -- constructed for
+              }
+  deriving (Show, Eq, Ord, Generic, Serialise)
+
+data ProofElem a = ProofElem
+  { nodeRoot    :: SizedMerkleRoot a
+  , siblingRoot :: SizedMerkleRoot a
+  , nodeSide    :: Side
+  } deriving (Show, Eq, Ord, Generic, Serialise)
+
+data Side = L | R
+  deriving (Show, Eq, Ord, Generic, Serialise)
+
+-- | Construct merkle proof by recusive walking the tree and collecting ProofElem until we hit
+-- matching leafRoot
+mkMerkleProof :: forall a. SizedMerkleTree a -- ^ merkle tree we want to
+                                             -- construct a proof from
+                        -> LeafIndex         -- ^ leaf index used for proof
+                        -> MerkleProof a
+mkMerkleProof SizedMerkleEmpty _ = MerkleProof [] Nothing
+mkMerkleProof (SizedMerkleTree rootNode) n =
+  let (path, proofLeaf) = constructPath [] rootNode
+  in MerkleProof path proofLeaf
+  where
+    constructPath :: [ProofElem a] -> SizedMerkleNode a -> ([ProofElem a], Maybe (SizedMerkleRoot a))
+    constructPath pElems (SizedMerkleLeaf leafRoot' leafIndex _)
+      | n == leafIndex = (pElems, Just leafRoot')
+      | otherwise = ([], Nothing)
+    constructPath pElems (SizedMerkleBranch bRoot ln rn) =
+         case (lPath, rPath) of
+              ((xs, Nothing), (ys, Nothing)) -> (xs ++ ys, Nothing)
+              ((xs, Just l), (ys, _))        -> (xs ++ ys, Just l)
+              ((xs, _), (ys, Just r))        -> (xs ++ ys, Just r)
+      where
+        lProofElem = ProofElem (mRoot ln) (mRoot rn) L
+        rProofElem = ProofElem (mRoot rn) (mRoot ln) R
+
+        lPath = constructPath (lProofElem:pElems) ln
+        rPath = constructPath (rProofElem:pElems) rn
+
+-- | Validate a merkle tree proof
+validateMerkleProof :: forall a. MerkleProof a -> SizedMerkleTree a -> Bool
```

Comment:
Why do we need whole tree to validate a proof? It should be validatable using only so called "root". In fact, the single thing you do with the tree is taking that root.

Validation is done on the client side, and the client doesn't have the tree at all; however she should be able to perform validation nonetheless.

Reply:
You are right, I will change this to `SizedMerkleRoot a`